### DISCUSSION TITLE
MacOS application package

### DISCUSSION THREAD

**Discussion by rafuck** [elfmz/far2l#881] [Ideas]
Hi,
I wrote simple script to packaging far2l to Far.app package for MacOS
https://github.com/rafuck/far2app
maybe someone will be useful.

**Comment by singalen** [comment]
In fact, the app package with all the plugins, scripts and the icon already exists in `<build>/install/` directory.

The only thing remaining is to make the app runnable on the hosts other then the build host.
This means:
* To add the non-system dylibs and frameworks to the app package, and change the `rpath` they are loaded from in `far2l` binaries.
* Probably to lower the compatible OS version with `-mmacosx-version-min`.

I have this script, but never got to fix all the TODOs there:

```
#!/bin/bash

set -euo pipefail

relocate_dep() {
    DEP="$1"
    BINARY="$2"
    LOADER_REL="${3:-}"
    echo "relocate_dep $1, $2"
    cp "${DEP}" far2l.app/Contents/MacOS/
    BASE_DEP=$(basename "${DEP}")
    install_name_tool -change "${DEP}" "@loader_path/${LOADER_REL}${BASE_DEP}" "${BINARY}"
}

package_deps() {
    echo "package_deps $1"
    BINARY="$1"
    LOADER_REL="${2:-}"

    DEPS=$(otool -L "${BINARY}" \
            | grep "/opt/local/lib" \
            | tr -s ' ' \
            | cut -d ' ' -f 1) || true # Failed grep is not a problem

    for DEP in ${DEPS}; do
        relocate_dep "${DEP}" "${BINARY}" "${LOADER_REL}"
    done

    echo "package_deps done with ${BINARY}:"
    otool -L "${BINARY}"
}

rm -rf ./far2l.app
cp -rf far2l/cmake-build-debug/install/far2l.app .

package_deps far2l.app/Contents/MacOS/far2l
for PLUG in far2l.app/Contents/MacOS/Plugins/*/plug/*.far-plug*; do
    echo PLUG="${PLUG}"
    package_deps "${PLUG}" "../../../"
done

# FIXME (done?): libarchive.dylib - copy one from /opt/local (the one we linked with)

# FIXME: Correctly process dylibs recursively.
# FIXME: Also add frameworks to the package (Wx, namely)
echo "Dylibs to process: $(ls -la far2l.app/Contents/MacOS/*.dylib)"
for DYLIB in far2l.app/Contents/MacOS/*.dylib; do
    echo DYLIB="${DYLIB}"
    package_deps "${DYLIB}"
done
```

**Reply by rafuck** [reply]
Hm... Thank you! But it means, that if I want far2l.app I must build far2l, and should not use HomeBrew?

**Reply by singalen** [reply]
You totally can use Homebrew, but it won’t run on the computers that don’t have it. 
And if somebody adds a Wx framework to this script (maybe I will), we will solve that problem as well.
The last issue with it is to package library licenses with them, to make it completely legally redistributable.

**Reply by rafuck** [reply]
It is not important may be, but when I tried create package, I find a problem: far2l executable must be located inside "bin" folder:
https://github.com/elfmz/far2l/blob/e65a978e91b29bb50a5f019c1d4f92a54d37dcdd/far2l/main.cpp#L379
it looks as a design mistake...

**Reply by singalen** [reply]
It's not that it *must*, it just *can*. This is for a case when the executable is installed into `/usr/bin` or `/usr/local/bin`, or `/opt/local/bin`, classic Unix-style.
MacOS can do this too, but application package is a more user friendly approach.
I suggest to simply ignore this code. I've been happily using the `.app` package for years now, and never had problems with it.
